components/homme/src/share/gllfvremap_mod.F90

```diff
@@ -0,0 +1,2253 @@
+
+  implicit none
+
+  private
+
+  ! Main API.
+  public :: &
+       ! Initialize this module.
+       gfr_init, &
+       ! Clean up this module.
+       gfr_finish, &
+       ! Remap GLL state to FV grid.
+       gfr_dyn_to_fv_phys, &
+       ! Remap FV tendencies or state to GLL grid.
+       gfr_fv_phys_to_dyn, &
+       ! Remap phis.
+       gfr_dyn_to_fv_phys_topo, &
+       gfr_fv_phys_to_dyn_topo, &
+       ! If nphys == 1, reconstruct the field to boost the OOA. If
+       ! nphys > 1, returns immediately.
+       gfr_pg1_reconstruct_topo, & ! call after the gfr_fv_phys_to_dyn_topo and the DSS
+       gfr_pg1_reconstruct         ! call after gfr_fv_phys_to_dyn and the DSS
+
+  integer, parameter :: nphys_max = np
+
+  real(kind=real_kind), parameter :: &
+       zero = 0.0_real_kind, half = 0.5_real_kind, &
+       one = 1.0_real_kind, two = 2.0_real_kind, &
+       four = 4.0_real_kind, eps = epsilon(1.0_real_kind)
+
+  ! Type for special case of pg1.
+  type, private :: Pg1SolverData_t
+     real(kind=real_kind) :: Achol(np*np,np*np), B(4,np*np), s(np*np), sts
+  end type Pg1SolverData_t
+
+  ! Top-level data type and functions for high-order, shape-preserving FV <->
+  ! GLL remap.
+  type, private :: GllFvRemap_t
+     integer :: nphys, npi
+     logical :: check, have_fv_topo_file_phis, boost_pg1
+     real(kind=real_kind) :: tolfac ! for checking
+     real(kind=real_kind) :: &
+          ! Node or cell weights
+          w_gg(np,np), &   ! GLL np
+          w_ff(nphys_max*nphys_max), & ! FV nphys
+          w_sgsg(np,np), & ! GLL npi
+          ! Mixed mass matrices
+          M_sgf(np,np,nphys_max,nphys_max), & ! GLL npi, FV nphys
+          M_gf(np,np,nphys_max,nphys_max), &  ! GLL np,  FV nphys
+          ! Interpolate from GLL npi to GLL np
+          interp(np,np,np,np), &
+          ! Remap FV nphys <-> GLL np
+          g2f_remapd(np,np,nphys_max*nphys_max), &
+          f2g_remapd(nphys_max*nphys_max,np,np)
+     ! FV subcell areas; FV analogue of GLL elem(ie)%metdet arrays
+     real(kind=real_kind), allocatable :: &
+          fv_metdet(:,:), & ! (nphys*nphys,nelemd)
+          ! Vector on ref elem -> vector on sphere
+          D_f(:,:,:,:,:), &   ! (nphys,nphys,2,2,nelemd)
+          ! Inverse of D_f
+          Dinv_f(:,:,:,:,:), &
+          qmin(:,:,:), qmax(:,:,:), &
+          phis(:,:)
+     type (spherical_polar_t), allocatable :: &
+          spherep_f(:,:,:) ! (nphys,nphys,nelemd)
+     type (Pg1SolverData_t) :: pg1sd
+  end type GllFvRemap_t
+
+  type (GllFvRemap_t), private :: gfr
+
+  ! For testing.
+  public :: &
+       gfr_test, &
+       gfr_g2f_scalar, gfr_f2g_scalar, gfr_f_get_latlon, gfr_f_get_cartesian3d, &
+       gfr_g_make_nonnegative, gfr_dyn_to_fv_phys_topo_elem, gfr_f2g_dss
+
+  ! Interfaces to support calling inside or outside a horizontally
+  ! threaded region.
+  interface gfr_dyn_to_fv_phys
+     module procedure gfr_dyn_to_fv_phys_hybrid
```

Comment:
is _hybrid for threaded? what is _dom_mt for? _mt probably stands multithreaded?

Reply:
dom_mt is the name of the structure that contains threading information. _hybrid is the impl that is called when in a threaded region. _dom_mt is the impl that is called when not, and it then sets up a threaded region (using dom_mt) and calls _hybrid.